### ISSUE TITLE
Unable to retrieve deal from storage provider

### ISSUE THREAD

**Issue by stuberman** [filecoin-project/boost#752]
### Checklist

- [X] This is **not** a question or a support request. If you have any boost related questions, please ask in the [discussion forum](https://github.com/filecoin-project/boost/discussions).
- [X] This is **not** a new feature or enhancement request. If it is, please open a [new idea discussion](https://github.com/filecoin-project/boost/discussions/new?category=ideas) instead. New feature and enhancement requests would be entertained by the boost team after a thorough discussion only.
- [X] I **have** searched on the [issue tracker](https://github.com/filecoin-project/boost/issues) and the [discussion forum](https://github.com/filecoin-project/boost/discussions), and there is no existing related issue or discussion.
- [X] I am running the [`Latest release`](https://github.com/filecoin-project/boost/releases), or the most recent RC(release canadiate) for the upcoming release or the dev branch(master), or have an issue updating to any of these.
- [X] I did not make any code changes to boost.

### Boost component

- [X] boost daemon - storage providers
- [ ] boost client
- [ ] boost UI
- [X] boost data-transfer
- [ ] boost index-provider
- [ ] Other

### Boost Version

```text
boostd version 1.3.0+git.0dffd87
Lotus Daemon:  1.17.1-rc3+mainnet+git.079dab0d7+api1.5.0
Local: lotus-miner version 1.17.1-rc3+mainnet+git.079dab0d7
```

### Describe the Bug

When attempting to retrieve a deal from SP (f01278) data transfer remains in `ResponderPaused` and deal will not transfer.
I was able to use booster-http to retrieve deal locally (see below)

### Logging Information

```text
boostd retrieval-deals list output

12D3KooWMgo5g34tVxk2ghd2TMkNS8wpTYJ62Rz2GLZTXcu2Vogc  1660123960946646043  ...sVz9zEL6  DealStatusErrored    0     262144       deal data transfer failed: data transfer channel 12D3KooWMgo5g34tVxk2ghd2TMkNS8wpTYJ62Rz2GLZTXcu2Vogc-12D3KooWSsaFCtzDJUEhLQYDdwoFtdCMqqfk562UMvccFz12kYxU-1660123960944002226 failed to transfer data: graphsync response to peer 12D3KooWMgo5g34tVxk2ghd2TMkNS8wpTYJ62Rz2GLZTXcu2Vogc did not complete: response status code RequestFailedUnknown
```

### Repo Steps

Issue noticed when @hannahhoward attempted to retrieve a deal from f01278:

`boostd data-transfers list -v`

> Sending Channels
> 
> ID          Status      Sending To                      Root Cid                    Initiated? Transferred Voucher                                                                                                                             
> 1661494380561613205 ResponderPaused 12D3KooWPmm3iECH2vjdezfxwKeEbHJ5RmcyT3Bz2zsPZmNZURiS QmWao9t1wBkQ54T2q4TiiAZfcNNfgwAAxfQK1NsVz9zEL6 N      0B      {"PayloadCID":{"/":"QmWao9t1wBkQ54T2q4TiiAZfcNNfgwAAxfQK1NsVz9zEL6"},"ID":533166,"Selector":{"Raw":"oWFSomFsoWRub25loGI6PqFhYaFhPqFhQKA="},"PieceCID":null,"PricePerByte":"0","PaymentInterval":10485760,"PaymentIntervalIncrease":10485760,"UnsealPrice":"0"}  
>  Message: stream reset

Using Boost Inspector for `QmWao9t1wBkQ54T2q4TiiAZfcNNfgwAAxfQK1NsVz9zEL6` yielded no results.
After retrieving over HTTP, Inspector still yielded no results.

We were able to use booster-http to retrieve deal:

`curl http://localhost:7777/payload/QmWao9t1wBkQ54T2q4TiiAZfcNNfgwAAxfQK1NsVz9zEL6 > /market/download.piece`

 | % Total | % Received | % Xferd | Average Speed | Time | Time | Time | Current | Dload | Upload | Total Spent | Left  Speed  | 
 | -  | -  | -  | -  | -  | -  | -  | -  | - | -  | -  | - |
 | 100 | 31.7G | 100 | 31.7G | 0 | 0 | 110M | 0 | 0:04:54 | 0:04:54 | --:--:-- | 110M | 

`boostx commp /market/download.piece`

> CommP CID:  baga6ea4seaqbkguoaif4qvdzma6syixru56wb3b72uwiohgq3uebemwwnlwocla
> Piece size:  34359738368
> Car file size:  34091302912

`boostd pieces piece-info baga6ea4seaqbkguoaif4qvdzma6syixru56wb3b72uwiohgq3uebemwwnlwocla`

> Piece:  baga6ea4seaqbkguoaif4qvdzma6syixru56wb3b72uwiohgq3uebemwwnlwocla
> Deals:

| DealID | SectorID |  Length | Offset | 
| - | - | - | - | 
| 2757338 | 2579 | 34359738368 | 0  | 

`boostd dagstore register-shard baga6ea4seaqbkguoaif4qvdzma6syixru56wb3b72uwiohgq3uebemwwnlwocla`

> Registered shard baga6ea4seaqbkguoaif4qvdzma6syixru56wb3b72uwiohgq3uebemwwnlwocla

`boostd dagstore list-shards | grep baga6ea4seaqbkguoaif4qvdzma6syixru56wb3b72uwiohgq3uebemwwnlwocla`

> baga6ea4seaqbkguoaif4qvdzma6syixru56wb3b72uwiohgq3uebemwwnlwocla  Available          

`lotus-miner storage find 2579`

> In d8cfac01-4999-45d5-8231-8aa9423224e2 (Unsealed, Sealed, Cache)
> 	Sealing: false; Storage: true
> 	Local (/pool/storage)
> 	URL: http://192.168.1.95:2345/remote/unsealed/s-t01278-2579

**Comment by LexLuthr** [comment]
@hannahhoward Can you take a look at this? That payload cannot be retrieved with graphsync but can be retrieved over http.

**Comment by LexLuthr** [comment]
I get below error when trying to retrieve the payload. 

```
# lotus client retrieve --miner f01278 QmWao9t1wBkQ54T2q4TiiAZfcNNfgwAAxfQK1NsVz9zEL6 /root/t
Recv 0 B, Paid 0 FIL, Open (New), 0s
Recv 0 B, Paid 0 FIL, DealProposed (WaitForAcceptance), 5ms
Recv 0 B, Paid 0 FIL, DealAccepted (Accepted), 355ms
Recv 0 B, Paid 0 FIL, PaymentChannelSkip (Ongoing), 355ms
Recv 256 KiB, Paid 0 FIL, BlocksReceived (Ongoing), 515ms
Recv 256 KiB, Paid 0 FIL, DataTransferError (Erroring), 517ms
Recv 256 KiB, Paid 0 FIL, BlockstoreFinalized (Errored), 520ms
ERROR: Retrieval Error: error generated by data transfer: deal data transfer failed: data transfer channel 12D3KooWMgo5g34tVxk2ghd2TMkNS8wpTYJ62Rz2GLZTXcu2Vogc-12D3KooWSsaFCtzDJUEhLQYDdwoFtdCMqqfk562UMvccFz12kYxU-1660123960944002227 failed to transfer data: channel 12D3KooWMgo5g34tVxk2ghd2TMkNS8wpTYJ62Rz2GLZTXcu2Vogc-12D3KooWSsaFCtzDJUEhLQYDdwoFtdCMqqfk562UMvccFz12kYxU-1660123960944002227: graphsync request failed to complete: request failed - unknown reason
```

**Comment by stuberman** [comment]
<div class="title" style="color: rgb(166, 30, 77); font-size: 1.25em; margin: 1em auto; text-align: center;">Deal<span> </span>bafyreiamewbr6bzip2g37ttnn7ptuxr4lggpgo6lmwe37j7xxlrgknvx7a</div><table class="deal-fields" style="border-collapse: collapse; padding: 0px; font-size: 1em; margin-bottom: 1em;"><tbody><tr><th style="color: rgb(59, 80, 108); font-weight: 400; padding: 0.5em 1em; text-align: left;">CreatedAt</th><td style="padding: 0.5em 1em; line-break: anywhere;">2021-11-14 17:53:54.195 <span class="aux" style="color: rgb(153, 153, 153);">(10 months<span> </span>ago)</span></td></tr><tr><th style="color: rgb(59, 80, 108); font-weight: 400; padding: 0.5em 1em; text-align: left;">Client Address</th><td style="padding: 0.5em 1em; line-break: anywhere;"><a href="https://filfox.info/en/address/f144zep4gitj73rrujd3jw6iprljicx6vl4wbeavi" target="_blank" rel="noreferrer" style="color: rgb(70, 82, 152); text-decoration: none;">f144zep4gitj73rrujd3jw6iprljicx6vl4wbeavi</a></td></tr><tr><th style="color: rgb(59, 80, 108); font-weight: 400; padding: 0.5em 1em; text-align: left;">Client Peer ID</th><td style="padding: 0.5em 1em; line-break: anywhere;">QmR4yWznXTQL2Weyzb7d1PgXxyERRjF6M2pNpCPWqyBWhD</td></tr><tr><th style="color: rgb(59, 80, 108); font-weight: 400; padding: 0.5em 1em; text-align: left;">Deal Data Root CID</th><td style="padding: 0.5em 1em; line-break: anywhere;"><a href="http://192.168.1.98:8080/inspect/bafybeifqi4yp5e6c4btxoon6n4dgac7lwds56lc6ra4tux4t2x2hsusxgm" style="color: rgb(70, 82, 152); text-decoration: none;">bafybeifqi4yp5e6c4btxoon6n4dgac7lwds56lc6ra4tux4t2x2hsusxgm</a></td></tr><tr><th style="color: rgb(59, 80, 108); font-weight: 400; padding: 0.5em 1em; text-align: left;">CAR File Path</th><td style="padding: 0.5em 1em; line-break: anywhere;"></td></tr><tr><th style="color: rgb(59, 80, 108); font-weight: 400; padding: 0.5em 1em; text-align: left;">Piece CID</th><td style="padding: 0.5em 1em; line-break: anywhere;"><a href="http://192.168.1.98:8080/inspect/baga6ea4seaqbkguoaif4qvdzma6syixru56wb3b72uwiohgq3uebemwwnlwocla" style="color: rgb(70, 82, 152); text-decoration: none;">baga6ea4seaqbkguoaif4qvdzma6syixru56wb3b72uwiohgq3uebemwwnlwocla</a></td></tr><tr><th style="color: rgb(59, 80, 108); font-weight: 400; padding: 0.5em 1em; text-align: left;">Verified</th><td style="padding: 0.5em 1em; line-break: anywhere;">No</td></tr><tr><th style="color: rgb(59, 80, 108); font-weight: 400; padding: 0.5em 1em; text-align: left;">Piece Size</th><td style="padding: 0.5em 1em; line-break: anywhere;">32 GiB <span class="aux" style="color: rgb(153, 153, 153);">(34,359,738,368<span> </span>bytes)</span></td></tr><tr><th style="color: rgb(59, 80, 108); font-weight: 400; padding: 0.5em 1em; text-align: left;">Piece File Path</th><td style="padding: 0.5em 1em; line-break: anywhere;"></td></tr><tr><th style="color: rgb(59, 80, 108); font-weight: 400; padding: 0.5em 1em; text-align: left;">Provider Collateral</th><td style="padding: 0.5em 1em; line-break: anywhere;">0.005 FIL</td></tr><tr><th style="color: rgb(59, 80, 108); font-weight: 400; padding: 0.5em 1em; text-align: left;">Funds Reserved</th><td style="padding: 0.5em 1em; line-break: anywhere;">0 atto</td></tr><tr><th style="color: rgb(59, 80, 108); font-weight: 400; padding: 0.5em 1em; text-align: left;">Current Epoch</th><td style="padding: 0.5em 1em; line-break: anywhere;">2,122,009</td></tr><tr><th style="color: rgb(59, 80, 108); font-weight: 400; padding: 0.5em 1em; text-align: left;">Start Epoch</th><td style="padding: 0.5em 1em; line-break: anywhere;">1,290,766<span class="aux" style="color: rgb(153, 153, 153);"><span> </span>(10 months)</span></td></tr><tr><th style="color: rgb(59, 80, 108); font-weight: 400; padding: 0.5em 1em; text-align: left;">End Epoch</th><td style="padding: 0.5em 1em; line-break: anywhere;">2,759,566<span class="aux" style="color: rgb(153, 153, 153);"><span> </span>(in 7 months)</span></td></tr><tr><th style="color: rgb(59, 80, 108); font-weight: 400; padding: 0.5em 1em; text-align: left;">Transfer Type</th><td style="padding: 0.5em 1em; line-break: anywhere;">manual</td></tr><tr><th style="color: rgb(59, 80, 108); font-weight: 400; padding: 0.5em 1em; text-align: left;">Transferred</th><td style="padding: 0.5em 1em; line-break: anywhere;">0 B <span class="aux" style="color: rgb(153, 153, 153);">(0<span> </span>bytes)</span></td></tr><tr><th style="color: rgb(59, 80, 108); font-weight: 400; padding: 0.5em 1em; text-align: left;">Transfer Size</th><td style="padding: 0.5em 1em; line-break: anywhere;"></td></tr><tr><th style="color: rgb(59, 80, 108); font-weight: 400; padding: 0.5em 1em; text-align: left;">Available for Retrieval</th><td style="padding: 0.5em 1em; line-break: anywhere;">Yes</td></tr><tr><th style="color: rgb(59, 80, 108); font-weight: 400; padding: 0.5em 1em; text-align: left;">Chain Deal ID</th><td style="padding: 0.5em 1em; line-break: anywhere;">2757338</td></tr><tr><th style="color: rgb(59, 80, 108); font-weight: 400; padding: 0.5em 1em; text-align: left;">Publish Message CID</th><td style="padding: 0.5em 1em; line-break: anywhere;"><a href="https://filfox.info/en/message/bafy2bzacedvvcmq7qnx653qg7g4eugatllotqpr4ker6jpuwxc54j3xcuj2aq" target="_blank" rel="noreferrer" style="color: rgb(70, 82, 152); text-decoration: none;">bafy2bzacedvvcmq7qnx653qg7g4eugatllotqpr4ker6jpuwxc54j3xcuj2aq</a></td></tr><tr><th style="color: rgb(59, 80, 108); font-weight: 400; padding: 0.5em 1em; text-align: left;">Status</th><td style="padding: 0.5em 1em; line-break: anywhere;">StorageDealActive</td></tr><tr><th style="color: rgb(59, 80, 108); font-weight: 400; padding: 0.5em 1em; text-align: left; font-family: &quot;Cerebri Sans&quot;, sans-serif; font-size: 16px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Message</th><br class="Apple-interchange-newline">Deal bafyreiamewbr6bzip2g37ttnn7ptuxr4lggpgo6lmwe37j7xxlrgknvx7a
CreatedAt	2021-11-14 17:53:54.195 (10 months ago)
Client Address	[f144zep4gitj73rrujd3jw6iprljicx6vl4wbeavi](https://filfox.info/en/address/f144zep4gitj73rrujd3jw6iprljicx6vl4wbeavi)
Client Peer ID	QmR4yWznXTQL2Weyzb7d1PgXxyERRjF6M2pNpCPWqyBWhD
Deal Data Root CID	[bafybeifqi4yp5e6c4btxoon6n4dgac7lwds56lc6ra4tux4t2x2hsusxgm](http://192.168.1.98:8080/inspect/bafybeifqi4yp5e6c4btxoon6n4dgac7lwds56lc6ra4tux4t2x2hsusxgm)
CAR File Path	
Piece CID	[baga6ea4seaqbkguoaif4qvdzma6syixru56wb3b72uwiohgq3uebemwwnlwocla](http://192.168.1.98:8080/inspect/baga6ea4seaqbkguoaif4qvdzma6syixru56wb3b72uwiohgq3uebemwwnlwocla)
Verified	No
Piece Size	32 GiB (34,359,738,368 bytes)
Piece File Path	
Provider Collateral	0.005 FIL
Funds Reserved	0 atto
Current Epoch	2,122,009
Start Epoch	1,290,766 (10 months)
End Epoch	2,759,566 (in 7 months)
Transfer Type	manual
Transferred	0 B (0 bytes)
Transfer Size	
Available for Retrieval	Yes
Chain Deal ID	2757338
Publish Message CID	[bafy2bzacedvvcmq7qnx653qg7g4eugatllotqpr4ker6jpuwxc54j3xcuj2aq](https://filfox.info/en/message/bafy2bzacedvvcmq7qnx653qg7g4eugatllotqpr4ker6jpuwxc54j3xcuj2aq)
Status	StorageDealActive
Message

**Comment by hannahhoward** [comment]
thanks @stuberman and @LexLuthr this is helpful and I will dig into it.

**Comment by hannahhoward** [comment]
@stuberman et all.

I believe I've gotten to the bottom of this one. Here's what is up:

The CID (`QmWao9t1wBkQ54T2q4TiiAZfcNNfgwAAxfQK1NsVz9zEL6 ) in question refers to a dag protobuf block that is malformed.

First, something I noticed from @stuberman's output is the CID in question is different from the root of the deal, which is `bafybeifqi4yp5e6c4btxoon6n4dgac7lwds56lc6ra4tux4t2x2hsusxgm`. I initiated a retrieval deal for this CID, and was able to download about 4 gigs, but the deal in question ultimately failed cause the CAR file stored is missing some of the child CIDs in the DAG that begins at the data root CID.

Anyway, undeterred, I went back and attempted a deal for the cid in question using filc. On the first run, it failed just as @LexLuthr 's had:

`graphsync request failed to complete: request failed - unknown reason`

However, on the second run, I got a different error message:

`graphsync request failed to complete: protobuf: (PBLink) invalid fieldNumber, expected 1, 2 or 3, got 13`

This caused me to go look at the blockstore for FilC, and I discovered on the first run, I had actually downloaded exactly one block -- in fact, I took the data in this block, and discovered that it hashes exactly to `QmWao9t1wBkQ54T2q4TiiAZfcNNfgwAAxfQK1NsVz9zEL6`

So, what happens here, on the first run, the remote reads the block data for `QmWao9t1wBkQ54T2q4TiiAZfcNNfgwAAxfQK1NsVz9zEL6` and sends it (hence, if you notice in @LexLuthr 's output it says `Recv 256 KiB, Paid 0 FIL` right before failing). However, because it's graphsync, the remote wants to send the rest of the DAG, so the remote attempts to decode the block and read its links... this is where the failure occurs. Graphsync only has response codes for errors (bad protocol design eek) so the code is RequestFailedUnknown.

On the second run, before contacting the remote, the local retrieval attempts to read it's local blocks to avoid downloading data it already has. It encounters the block, attempts to decode it, and fails, but this time locally, so it's able to report a better error.

The long and short here is the stored IPLD data has issues, and now if we want to trace this further, it would be best to track down the storer `QmR4yWznXTQL2Weyzb7d1PgXxyERRjF6M2pNpCPWqyBWhD` and figure out why there were issues in the data they put into the CAR file they stored.

In the meantime, I think we can close this issue.
